Add handling of tuples in type subscriptions

bskinn opened the pull request:
Fixes #162.

Initial implementation and tests added, and project lints fixed.

There are some significant questions to be resolved before merging. Prime among them is whether introducing `nptyping` as a test/dev dependency is a good idea.

Also, input on whether this implementation is sufficiently general, and whether additional/different test cases should be added/used would be good.

bskinn committed:
Add internal-tuple implementation

bskinn committed:
Add internal-tuple tests

bskinn committed:
Update project config

bskinn commented on src/sphinx_autodoc_typehints/__init__.py:
```diff
@@ -150,7 +169,10 @@ def format_annotation(annotation: Any, config: Config) -> str:
-        fmt = [format_annotation(arg, config) for arg in args]
+        try:
+            fmt = [format_annotation(arg, config) for arg in args]
+        except TypeError:
```
Also not entirely sure if this exception handling needs to be refined.

gaborbernat replied:
Why do we need this?

bskinn replied:
To handle things with bare integers in annotations, like `Float[32]` in `nptyping` and (presumably) [`Literal[640]` in PEP 646](https://www.python.org/dev/peps/pep-0646/#use-case-1-specifying-shape-values).

The `TypeError` here is raised because `args` is not iterable in these cases. Singletons are not an issue in other subscriptable types because they are special-cased in their handling earlier in the function.

I did the error trapping this way initially b/c EAFP model. I can think of two other ways to handle offhand at this point in the code, both of which are LBYL, but more precise:

1. Switch to a `try: iter(args); except ...` handler. Would make it more specifically catch the iterable/non-iterable difference.

2. Switch to an explicit `isinstance(args, Number)` check. This would be a very focused check, and would not cover cases with non-numeric singletons. I believe neither `nptyping` nor PEP 646 uses non-numeric singletons, but I'm not 100% sure; and, who knows what's out there in the ecosystem, or will be.

I think I'm inclined to revise to (1).

gaborbernat replied:
Sounds good for 1 👍

gaborbernat commented on src/sphinx_autodoc_typehints/__init__.py:
```diff
@@ -89,6 +89,22 @@ def get_annotation_args(annotation: Any, module: str, class_name: str) -> tuple[
 
 
+def format_internal_tuple(t: tuple, config: Config) -> str:
+    # An annotation can be a tuple, e.g., for nptyping:
+    #   NDArray[(typing.Any, ...), Float]
+    # In this case, format_annotation receives:
+    #   (typing.Any, Ellipsis)
+    # This solution should hopefully be general for *any* type that allows
+    # tuples in annotations
```
Please align it alongside 120 characters

bskinn replied:
I'm not sure what you mean -- do you want me to remove the indentation?

gaborbernat replied:
yes; this line could reach all the way to 120 character, so no need to break it so soon

bskinn replied:
Fixed in aa53976

gaborbernat commented:
> Prime among them is whether introducing nptyping as a test/dev dependency is a good idea.

I see it's a pure wheel so I'm ok with this.

> Also, input on whether this implementation is sufficiently general, and whether additional/different test cases should be added/used would be good.

Can be a good starting place. We can add additional tests if users report bugs, no?

bskinn commented:
> > Prime among them is whether introducing nptyping as a test/dev dependency is a good idea.
> 
> I see it's a pure wheel so I'm ok with this.
> 

The main thing I was thinking of is, currently my test cases depend on some internal type transformations that `nptyping` does. If `nptyping` changes these transformations, that will cause this project's test suite to fail. I realized: I can just change the test examples so that `nptyping` does not transform them. That should remove the brittleness, at least most of the way. I'll make this change.

> > Also, input on whether this implementation is sufficiently general, and whether additional/different test cases should be added/used would be good.
> 
> Can be a good starting place. We can add additional tests if users report bugs, no?

Yep. That's what I was thinking.

gaborbernat commented:
> That should remove the brittleness, at least most of the way. I'll make this change.

Sounds good on this.

bskinn committed:
De-fold comment to one line

https://github.com/tox-dev/sphinx-autodoc-typehints/pull/212/files/d3ea411f39ae55aa683bcb15edd78e0078ae2553#r789741412

bskinn commented:
For my reference:

- [ ] De-fold comment
- [ ] Re-work iterable/non-iterable check
- [ ] Fix typing problem
- [ ] Rewrite tests to avoid `nptyping` brittleness

bskinn committed:
Fix/silence mypy errors

Added tuple subscription to helper function.

Interestingly, despite it being a typing package, nptyping isn't
compatible yet with static typecheckers, and thus doesn't
have typing in its source, nor does it have a .pyi (AFAICT).

https://github.com/ramonhagenaars/nptyping/issues/34

The nptyping annotations are still helpful for those reading
the code, though.

bskinn committed:
Revise nptyping test cases to reduce brittleness

Some dependence on the internal module structure of the nptyping
package is unavoidable here, unfortunately, so the brittleness
cannot be completely removed.

bskinn commented:
b279611 removes as much of the brittleness from the `nptyping` dependency as is possible with the current [`assert result == expected_result` check](https://github.com/tox-dev/sphinx-autodoc-typehints/blob/7b8c357be15d7b7b4ec4ad272f554f7ab3e0e197/tests/test_sphinx_autodoc_typehints.py#L209).

If `nptyping` restructures/refactors their internals, these test cases will break.

It would be possible to reduce/eliminate this risk of breakage by loosening the assert, perhaps only for the `nptyping` cases; e.g.,:

```

if "nptyping" in result:
    # Looser check that matches e.g. re.search(expected_regex, result)
    # where expected_regex has things in it like r"~nptyping\..+\.NDArray"
else:
    assert result == expected_result
```

That would make the test function messy, though.

I guess another alternative would be to make a new test function for the nptyping cases, with the above selective-matching behavior.

-----

TLDR; are you okay with accepting this possibility for breakage from `nptyping` for now, and we'll see what happens? Or, would you prefer to try to mitigate it now?

bskinn committed:
Revise singleton parameter handling

Switch to LBYL approach by checking iter(args) and treating
as a singleton on TypeError, else as iterable.

gaborbernat reviewed (changes requested):
Please add a changelog with 1.16.0

https://github.com/tox-dev/sphinx-autodoc-typehints/blob/main/CHANGELOG.md

bskinn committed:
Add CHANGELOG entry

bskinn commented:
> Please add a changelog with 1.16.0
> 
> https://github.com/tox-dev/sphinx-autodoc-typehints/blob/main/CHANGELOG.md

Done!
